File: Assets/Scripts/Progression/SkillTree.cs
using System.Collections.Generic;
using UnityEngine;
using ActionGameDemo.Character;

namespace ActionGameDemo.Progression
{
    [CreateAssetMenu(fileName = "SkillTree", menuName = "ActionGameDemo/Skill Tree")]
    public class SkillTree : ScriptableObject
    {
        [SerializeField] private List<SkillTreeNodeData> _nodes = new List<SkillTreeNodeData>();
        private HashSet<string> _unlockedNodes = new HashSet<string>();
        private int _availableSkillPoints = 0;

        public List<SkillTreeNodeData> Nodes => _nodes;
        public int AvailableSkillPoints => _availableSkillPoints;
        public IReadOnlyCollection<string> UnlockedNodeIds => _unlockedNodes;

        public event System.Action<string> OnNodeUnlocked;
        public event System.Action<int> OnSkillPointsChanged;

        public void AddSkillPoints(int amount)
        {
            _availableSkillPoints += Mathf.Max(0, amount);
            OnSkillPointsChanged?.Invoke(_availableSkillPoints);
        }

        public bool CanUnlock(string nodeId, int playerLevel)
        {
            var node = FindNode(nodeId);
            if (node == null) return false;
            if (_unlockedNodes.Contains(nodeId)) return false;
            if (playerLevel < node.requiredLevel) return false;
            if (_availableSkillPoints < node.cost) return false;

            foreach (var prereqId in node.prerequisiteNodeIds)
            {
                if (!_unlockedNodes.Contains(prereqId)) return false;
            }
            return true;
        }

        public bool TryUnlock(string nodeId, int playerLevel, CharacterStats stats)
        {
            if (!CanUnlock(nodeId, playerLevel)) return false;

            var node = FindNode(nodeId);
            _availableSkillPoints -= node.cost;
            _unlockedNodes.Add(nodeId);
            ApplyNodeEffects(node, stats);

            OnNodeUnlocked?.Invoke(nodeId);
            OnSkillPointsChanged?.Invoke(_availableSkillPoints);
            return true;
        }

        public bool IsUnlocked(string nodeId) => _unlockedNodes.Contains(nodeId);

        private SkillTreeNodeData FindNode(string nodeId)
        {
            return _nodes.Find(n => n.nodeId == nodeId);
        }

        private void ApplyNodeEffects(SkillTreeNodeData node, CharacterStats stats)
        {
            if (stats == null) return;
            if (node.attackBonusPercent != 0f) stats.AddAttackModifier(node.attackBonusPercent / 100f);
            if (node.defenseBonusPercent != 0f) stats.AddDefenseModifier(node.defenseBonusPercent / 100f);
            if (node.criticalRateBonus != 0f) stats.AddCriticalRateModifier(node.criticalRateBonus / 100f);
            if (node.healthBonusPercent != 0f)
            {
                float bonus = stats.MaxHealth * node.healthBonusPercent / 100f;
                stats.SetMaxHealth(stats.MaxHealth + bonus);
                stats.ModifyHealth(bonus);
            }
        }

        public void ClearUnlocks()
        {
            _unlockedNodes.Clear();
            _availableSkillPoints = 0;
        }

        public List<string> GetUnlockedNodeIds() => new List<string>(_unlockedNodes);
    }
}

File: server/src/main/java/com/actiongame/server/domain/progression/SkillTree.java
package com.actiongame.server.domain.progression;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * 技能树 (服务端权威)
 * 对应Unity SkillTree, 纯数据逻辑
 */
public class SkillTree {
    private final List<SkillTreeNode> nodes;
    private final Set<String> unlockedNodes = new HashSet<>();
    private int availableSkillPoints = 0;

    public SkillTree(List<SkillTreeNode> nodes) {
        this.nodes = nodes;
    }

    public void addSkillPoints(int amount) {
        availableSkillPoints += Math.max(0, amount);
    }

    public boolean canUnlock(String nodeId, int playerLevel) {
        SkillTreeNode node = findNode(nodeId);
        if (node == null) return false;
        if (unlockedNodes.contains(nodeId)) return false;
        if (playerLevel < node.getRequiredLevel()) return false;
        if (availableSkillPoints < node.getCost()) return false;
        for (String prereqId : node.getPrerequisiteNodeIds()) {
            if (!unlockedNodes.contains(prereqId)) return false;
        }
        return true;
    }

    public boolean tryUnlock(String nodeId, int playerLevel) {
        if (!canUnlock(nodeId, playerLevel)) return false;
        SkillTreeNode node = findNode(nodeId);
        availableSkillPoints -= node.getCost();
        unlockedNodes.add(nodeId);
        return true;
    }

    public boolean isUnlocked(String nodeId) {
        return unlockedNodes.contains(nodeId);
    }

    public void clearUnlocks() {
        unlockedNodes.clear();
        availableSkillPoints = 0;
    }

    private SkillTreeNode findNode(String nodeId) {
        for (SkillTreeNode node : nodes) {
            if (node.getNodeId().equals(nodeId)) return node;
        }
        return null;
    }

    public List<SkillTreeNode> getNodes() { return nodes; }
    public int getAvailableSkillPoints() { return availableSkillPoints; }
    public Set<String> getUnlockedNodes() { return unlockedNodes; }
}
